app/services/atmosphere/billing_service.rb

```diff
@@ -20,45 +20,57 @@ def self.bill_all_appliances
-              Rails.logger.warn("The balance of fund #{appl.fund.id} is insufficient to cover continued operation of appliance #{appl.id} (current balance: #{appl.fund.balance}; overdraft limit: #{appl.fund.overdraft_limit}; calculated charge: #{amount_due}). Flagging appliance as expired.")
-              BillingLog.new(timestamp: Time.now.utc, user: appl.appliance_set.user, appliance: appl.id.to_s+'---'+(appl.name.blank? ? 'unnamed_appliance' : appl.name), fund: appl.fund.name, message: "Funding expired for this appliance.", actor: "bill_appliance", amount_billed: 0).save
-              appl.billing_state = "expired" # A separate method will be used to clean up all expired appliances according to their funds' termination policies
-              appl.save
+              # We've run out of funds. This deployment cannot be paid for. Flagging it as expired.
+              Rails.logger.warn("The balance of fund #{appl.fund.id} is insufficient to cover continued operation of deployment #{dep.id} belonging to appliance #{appl.id} (current balance: #{appl.fund.balance}; overdraft limit: #{appl.fund.overdraft_limit}; calculated charge: #{amount_due}). Flagging deployment as expired.")
+              BillingLog.new(timestamp: Time.now.utc, user: appl.appliance_set.user, appliance: appl.id.to_s+'---'+(appl.name.blank? ? 'unnamed_appliance' : appl.name), fund: appl.fund.name, message: "Funding expired for deployment #{dep.id}.", actor: "bill_appliance", amount_billed: 0).save
+              dep.billing_state = "expired" # A separate method will be used to clean up all expired deployments according to their funds' termination policies
+              dep.save
             else
-              Rails.logger.debug("Applying charge of #{amount_due} to appliance #{appl.id} and deducting it from balance of fund #{appl.fund.id}.")
+              Rails.logger.debug("Applying charge of #{amount_due} to deployment #{dep.id} belonging to appliance #{appl.id} and deducting it from balance of fund #{appl.fund.id}.")
               appl.fund.balance -= amount_due
               appl.fund.save
               appl.amount_billed += amount_due
-              appl.prepaid_until = billing_time+(apply_prepayment ? @appliance_prepayment_interval : 0)
               appl.last_billing = billing_time
-              appl.billing_state = "prepaid"
+              dep.prepaid_until = billing_time+(apply_prepayment ? @appliance_prepayment_interval : 0)
+              dep.billing_state = "prepaid"
               appl.save
-              if appl.errors.blank?
+              dep.save
+              if dep.errors.blank?
                 # Write success to log.
                 BillingLog.new(timestamp: Time.now.utc, user: appl.appliance_set.user, appliance: appl.id.to_s+'---'+(appl.name.blank? ? 'unnamed_appliance' : appl.name), fund: appl.fund.name, message: message, actor: "bill_appliance", amount_billed: amount_due).save
               else
```

Comment:
Good catch. I have modified the code to run appl.save and dep.save transactionally, and to update the log outside of the transaction, ensuring that (1) consistency is maintained; (2) log message is written regardless of outcome.